docs/man/fhs_makepreload.8.in
.\" fhs_makepreload - Fast Hyperspace tool to make a preload hash file (makepreload)
.TH "fhs_makepreload" "8" "Aug 2013"  "Trever Adams" ""
.SH "NAME"
fhs_makepreload \- Fast Hyperspace tool to make a preload hash file (makepreload)
.SH "SYNOPSIS"
\fBfhs_makepreload\fP -p \fIPRIMARY_HASH_SEED\fP -s \fISECONDARY_HASH_SEED\fP -d \fIFHS_DIRECTORY\FP -o \fIOUTPUT_FHS_PRELOAD_FILE\fP
.PP
.SH "DESCRIPTION"
.PP
\fBfhs_makepreload\fP is a command-line tool to make a preload hash for for fhs classifiers. fhs stands for Fast Hyperspace.
.PP

.PP
.SH "OPTIONS"
.PP
.BR PRIMARY_HASH_SEED
.br
.BR SECONDARY_HASH_SEED
.PP
   The hash seeds are used to prime the hashing algorithm. If
   these do not match in both the trainer, when training the
   data, and the classifier when classifying, the results are
   undefined. For this reason, this can be used as a sort of
   pseudo copy protection.
.PP
   Additionally, these values determine, to an extent, the
   number of collisions in hash data. The only way to know if
   you have chosen a good value is testing.
.PP
   Each seed is a hexadecimal representation of a 32 bit number.
   If is \fBnot\fR prefixed with "0x". All alpha characters
   should be uppercase.
.PP
   These \fBmust\fR be the same for all categories.
.PP
.BR FHS_DIRECTORY
.PP
   This is the directory where the fhs files, aka fhs
   classifier data, are stored. All the data here will
   be read in and a preload written out.
.PP
.BR OUTPUT_FHS_PRELOAD_FILE
.PP
   This is the name which the preload file will be given.
   It should not contain any directory compnents as the file
   will be written to the \fBFHS_DIRECTORY\fR.
.PP
WARNING: Spaces and case matter.
.PP
.SH "NOTES"
You must always have all the files within each category share the hash seeds.
Additionally, all categories within the same classifier data set \fBmust\fR
share the same hash seeds.
.PP
The use of a preload file dramatically decreases the speed of start-up time for the classifier!
.PP

.SH "FILES"
.nf
NONE
.fi

.PP
.SH "SEE ALSO"
.nf
.I fhs_findtolearn (8)
.I fhs_judge(8)
.I fhs_learn (8)
.fi

.PP
.SH "AUTHORS"
.nf
Trever Adams
.fi

.PP
.SH "BUGS"
There of course aren't any bugs, but if you find any, you should first
consult https://github.com/treveradams/C-ICAP-Classify and, if
necessary, file a bug report there.
.fi


services/classify/fhs_makepreload.c
/*
 *  Copyright (C) 2008-2012 Trever L. Adams
 *
 *  This file is part of srv_classify c-icap module and accompanying tools.
 *
 *  srv_classify is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU Lesser General Public License as
 *  published by the Free Software Foundation, either version 3 of
 *  the License, or (at your option) any later version.
 *
 *  srv_classify is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public License
 *  along with Foobar.  If not, see <http://www.gnu.org/licenses/>.
 */


#define _GNU_SOURCE

#ifndef _FILE_OFFSET_BITS
#define _FILE_OFFSET_BITS 64
#endif
#if (_FILE_OFFSET_BITS != 64)
#undef _FILE_OFFSET_BITS
#define _FILE_OFFSET_BITS 64
#endif

#define NOT_CICAP

#include <stdio.h>
#include <stdint.h>
#include <inttypes.h>
#include <limits.h>
#include <stdlib.h>
#include <string.h>
#include <locale.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <unistd.h>
#include <fcntl.h>
#include <wchar.h>
#include <wctype.h>
#include <time.h>
#include <sys/types.h>
#include <dirent.h>

#include "hash.c"
#include "hyperspace.c"

char *fhs_out_file;
char *fhs_dir;

int readArguments(int argc, char *argv[])
{
int i;
	if(argc < 9)
	{
		printf("Format of arguments is:\n");
		printf("\t-p PRIMARY_HASH_SEED\n");
		printf("\t-s SECONDARY_HASH_SEED\n");
		printf("\t-d FHS DIRECTORY TO READ\n");
		printf("\t-o OUTPUT_FHS_PRELOAD_FILE, must be a file name only, will be saved to FHS DIRECTORY\n");
		printf("Spaces and case matter.\n");
		return -1;
	}
	for(i=1; i<9; i+=2)
	{
		if(strcmp(argv[i], "-p") == 0) sscanf(argv[i+1], "%"PRIx32, &HASHSEED1);
		else if(strcmp(argv[i], "-s") == 0) sscanf(argv[i+1], "%"PRIx32, &HASHSEED2);
		else if(strcmp(argv[i], "-o") == 0)
		{
			fhs_out_file = malloc(strlen(argv[i+1]) + 1);
			sscanf(argv[i+1], "%s", fhs_out_file);
		}
		else if(strcmp(argv[i], "-d") == 0)
		{
			fhs_dir = malloc(strlen(argv[i+1]) + 1);
			sscanf(argv[i+1], "%s", fhs_dir);
		}
	}
/*	printf("Primary Seed: %"PRIX32"\n", HASHSEED1);
	printf("Secondary Seed: %"PRIX32"\n", HASHSEED2);
	printf("Learn File: %s\n", judge_file);*/
	return 0;
}

void loadMassCategories(void)
{
struct dirent *current_file = NULL;
DIR *directory;
	chdir(fhs_dir);

	directory=opendir(".");
	if(directory==NULL)
	{
		printf("Unable to open FHS Directory provided!\n");
		exit(-1);
	}
	do {
		current_file = readdir(directory);
		if (current_file!=NULL && strcmp(current_file->d_name, ".") != 0 && strcmp(current_file->d_name, "..") != 0 && strcmp(current_file->d_name, fhs_out_file) != 0 && strcmp(&current_file->d_name[strlen(current_file->d_name)-4], ".fhs") == 0)
		{
			printf("Loading FHS file: %s\n", current_file->d_name);
			loadHyperSpaceCategory(current_file->d_name, current_file->d_name);
		}
	} while (current_file != NULL);
	closedir(directory);
}

int main(int argc, char *argv[])
{
int fhs_file;
FHS_HEADERv1 header;
clock_t start, end;
uint_least16_t docsWritten = 0;
	initHTML();
	initHyperSpaceClassifier();
	if(readArguments(argc, argv) == -1) exit(-1);

	printf("Loading hashes -- be patient!\n");
	start=clock();
	loadMassCategories();
	printf("\nWriting out preload file: %s\n", fhs_out_file);

	fhs_file = openFHS(fhs_out_file, &header, 1);

	docsWritten = writeFHSHashesPreload(fhs_file, &header, &HSJudgeHashList);

	close(fhs_file);

	end=clock();
	printf("Wrote out: %"PRIu32" hashes as %"PRIu16" documents.\n", HSJudgeHashList.used, docsWritten);
	printf("Preload making took %lf seconds\n", (double)((end-start)/(CLOCKS_PER_SEC)));

	deinitHyperSpaceClassifier();
	deinitHTML();
	return 0;
}
